### ISSUE TITLE
在demo.py上改为视频检测

### ISSUE THREAD

**Issue by emojiLee** [zju3dv/pvnet#104]
思达你好，我有如下几个问题，目的：单张图片转实时视频拿小猫咪测试显示预测框。

1.请问tools/demo.py的demo()中如下这一行：
seg_pred, vertex_pred, loss_seg, loss_vertex, precision, recall = net(image, mask, vertex, vertex_weights)里面的net调用的__call__和forward在哪里？因为当我删除vertex_weights的时候，结果报错缺少位置参数。
2.我看到pvnet论文中的流程，输入img,经过resnet和上采样，求解出seg_pred, vertex_pred，这个过程在视频测试的时候，我得到实时的每一帧图片，还需要每一帧分割的mask.png？
3.相机的内置矩阵我是需要使用自己摄像头用标定板验证得到么？
4.LINEMOD数据集每个object里面都有一个corners.txt作为验证使用到的cls(cat)_bb8_3d.txt，里面是bounding box的8个坐标点，那这个corners.txt怎么得来的？

问题有点多，期待您的回复，感谢呀！

**Comment by pengsida** [comment]
1. https://github.com/zju3dv/pvnet/blob/master/tools/demo.py#L37
2. 如果想测试mask的准确度，需要mask.png。如果是想得到可视化结果，不需要。
3. 需要标定自己的摄像头。
4. https://github.com/zju3dv/pvnet/blob/master/lib/utils/data_utils.py#L72

**Comment by emojiLee** [comment]
真厉害！谢谢~
不好意思，我之前看的不仔细

**Comment by emojiLee** [comment]
思达，我有问题，我只输入rgb图片，用了如下几张图片，为什么我用自己拍摄的环境，猫咪图片不可以预测，加了标定板就好一些，而blender融合的图片没有标定板就能预测呢？是否验证需要标定板呢？相机标定如何确定得到的内参是准确呢？（我的做法：7*9的标定纸固定不变，相机不同角度横着拍摄了20张照片，代码执行出的相机内参）

1. LINMEMOD的任意一张图片。可预测
![Figure_444](https://user-images.githubusercontent.com/51335035/68195694-3fdb0e80-fff2-11e9-8f14-4cef0a6fae8f.png)

2.blender小猫猫的一张图片。可预测
![Figure_8879](https://user-images.githubusercontent.com/51335035/68195733-508b8480-fff2-11e9-81f9-27a87bb08815.png)

3.电脑摄像头拍的一张照片，不可以，坐标点很乱。（有用到自己标定的内参矩阵，但是不知道标定的如何才算准确）
![Figure_666](https://user-images.githubusercontent.com/51335035/68195767-68fb9f00-fff2-11e9-8f3d-0704feda1fc3.png)

4.把小猫猫放在标定纸上，再用相机拍摄，然后用了我的相机内参和原代码里的。
（1）用我的相机内参
![Figure_000](https://user-images.githubusercontent.com/51335035/68196187-2a1a1900-fff3-11e9-8b63-a54e422677e9.png)

（2）用原代码相机内参
![Figure_8](https://user-images.githubusercontent.com/51335035/68195794-7c0e6f00-fff2-11e9-986a-39ac1c3c5b38.png)

**Comment by pengsida** [comment]
1. 你可以可视化看关键点是不是正确的。
2. 标定板更好，是因为有标定板的图片背景更干净。

**Comment by emojiLee** [comment]
思达，嗯，我调摄像头实时检测了，但是不稳定，有没有什么建议使得更稳定一些呢？谢谢！
![lina](https://user-images.githubusercontent.com/51335035/68379691-54054400-0189-11ea-86cf-2a3e20cf3a7d.gif)

**Comment by pengsida** [comment]
建议使用一个detector，把图片中的物体crop出来，再做关键点检测。这样更快，效果更好。

**Comment by pengsida** [comment]
在自己打印的这个物体上finetune，效果更稳定。

**Comment by emojiLee** [comment]
嗯嗯，谢谢思达，真厉害

**Comment by naaths** [comment]
您好，能分享一下您的实时视频demo代码吗？
另外我也是在自己拍的照片里无法预测，您是怎么在哪里改的相机内参的，谢谢啦

**Comment by naaths** [comment]
您好，能分享一下您的实时视频demo代码吗？
另外我也是在自己拍的照片里无法预测，您是怎么在哪里改的相机内参的，谢谢啦

**Comment by emojiLee** [comment]
@naaths 

1. 以下是pvnet基本的实时检测代码，实时捕获的图像是640*480，如果你设置为其他长宽，模型处理和画框的代码自己修改下。
 `
import sys
sys.path.append('.')
sys.path.append('..')
from lib.networks.model_repository import *
from lib.utils.net_utils import load_model
import torch
from lib.ransac_voting_gpu_layer.ransac_voting_gpu import ransac_voting_layer_v3
from lib.utils.evaluation_utils import pnp
from lib.utils.base_utils import Projector
import json
from lib.utils.config import cfg
from torch.nn import DataParallel
from torch import nn, optim
import os
from PIL import Image
import numpy as np
np.set_printoptions(suppress=True)
import matplotlib.patches as patches
import cv2
import time

print(sys.path)
with open('configs/linemod_train.json', 'r') as f:
    train_cfg = json.load(f)
train_cfg['model_name'] = '{}_{}'.format('cat', train_cfg['model_name'])  # 训练后的模型保存在data/model/xxx_linemod_train中

vote_num = 9

class NetWrapper(nn.Module):
    def __init__(self, net):
        super(NetWrapper, self).__init__()
        self.net = net
        self.criterion = nn.CrossEntropyLoss(reduce=False)

    def forward(self, image):
        seg_pred, vertex_pred = self.net(image)
        return seg_pred, vertex_pred

class EvalWrapper(nn.Module):
    def forward(self, seg_pred, vertex_pred, use_argmax=True):
        vertex_pred = vertex_pred.permute(0, 2, 3, 1)
        b, h, w, vn_2 = vertex_pred.shape
        vertex_pred = vertex_pred.view(b, h, w, vn_2 // 2, 2)
        if use_argmax:
            mask = torch.argmax(seg_pred, 1)
        else:
            mask = seg_pred
        s = ransac_voting_layer_v3(mask, vertex_pred, 512, inlier_thresh=0.99)
        #print(('lixiaohong test s_ransac_voting_layer_v3:'),s)
        return s

def read_data(origin_rgb):
    import torchvision.transforms as transforms

    demo_dir_path = os.path.join(cfg.DATA_DIR, 'cat_demo')

    points_3d = np.loadtxt(os.path.join(demo_dir_path, 'cat_points_3d.txt'))
    bb8_3d = np.loadtxt(os.path.join(demo_dir_path, 'cat_bb8_3d.txt'))

    projector = Projector()

    transformer = transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize(mean=[0.485, 0.456, 0.406],
                             std=[0.229, 0.224, 0.225])
    ])

    rgb = transformer(origin_rgb)
    return rgb, points_3d, bb8_3d

def demo():
    net = Resnet18_8s(ver_dim=vote_num * 2, seg_dim=2)
    net = NetWrapper(net).cuda()
    net = DataParallel(net)

    optimizer = optim.Adam(net.parameters(), lr=train_cfg['lr'])
    model_dir = os.path.join(cfg.MODEL_DIR, "cat_linemod_train")
    load_model(net.module.net, optimizer, model_dir, -1) 

    vid = cv2.VideoCapture(0)
    vid.set(3, 640)
    vid.set(4, 480)
    while (vid.isOpened()):
        i = 1
        ret, rgb = vid.read()
        start_time = time.time()
        image, points_3d, bb8_3d = read_data(rgb)
        image = image.unsqueeze(0).cuda()
        seg_pred, vertex_pred = net(image)
        eval_net = DataParallel(EvalWrapper().cuda())
        corner_pred = eval_net(seg_pred, vertex_pred).cpu().detach().numpy()[0]
        print('lxh corner_pred: ', corner_pred)
		
        # 相机内参矩阵
        camera_matrix = np.array([[572.4114, 0., 325.2611],
			              [0., 573.57043, 242.04899],
			              [0., 0., 1.]])

        pose_pred = pnp(points_3d, corner_pred, camera_matrix)
        print(('预测的rt:'),pose_pred)
        projector = Projector()
        bb8_2d_pred = projector.project(bb8_3d, pose_pred, 'linemod')
        print(('640*480的2D角点值:'),bb8_2d_pred)

        # draw
        xy1 = bb8_2d_pred[[0, 1, 3, 2, 0, 4, 6, 2]]
        xy2 = bb8_2d_pred[[5, 4, 6, 7, 5, 1, 3, 7]]
        cv2.polylines(rgb, np.int32([xy1]), True, (0, 0, 255), 2)
        cv2.polylines(rgb, np.int32([xy2]), True, (0, 0, 255), 2)
        end_time = time.time()
        cv2.putText(rgb, '{:.2f}ms'.format((end_time - start_time) * 1000), (40, 40), 0,
                    fontScale=1, color=(0, 0, 0), thickness=2)

        cv2.imshow('Image', rgb)
        if cv2.waitKey(1) & 0xFF == ord('q'):
            break

    vid.release()
    cv2.destroyAllWindows() 

if __name__ == "__main__":
    demo()
`
备注：代码如果出错，需要注意的地方。
<1>修改configs/linemod_train.json路径或其他文件路径。
<2>修改lib/ransac_voting_gpu_layer/ransac_voting_gpu.py文件中def b_inv(b_mat)函数。（具体按自己想法修改）

2. 
（1）linemod数据训模，拿小猫进行测试，自己拍摄的照片或实时检测时摄像头距离小猫要半米以上测试为好。要是自己的数据集，按自身情况。
（2）linemod数据测试时不需要更改内参，自己的数据测试时修改上面代码的camera_matrix （3*3矩阵）为自己定义的相机内参即可。

**Comment by pengsida** [comment]
@emojiLee 你太酷了

**Comment by emojiLee** [comment]
> @emojiLee 你太酷了
哈哈，一直关注issue，看有人问来着

**Comment by coocooon** [comment]
@emojiLee 谢谢哈,但是还有点小问题想请教一下,我可视化了关键点,发现关键点正确,但是包围盒不对,自己的相机参数和源代码的都试了,都不行,在custom中正常.这是什么原因
![2020-10-23 19-15-04 的屏幕截图](https://user-images.githubusercontent.com/64255092/96997920-40ebaf80-1565-11eb-837f-00d533d3649c.png)

**Comment by naaths** [comment]
@emojiLee 您好，在您的实时代码中，cat_points_3d.txt和cat_bb8_3d.txt是官方的吗，我怎么没找到，能发一下吗，qq邮箱：196715133@qq.com
谢谢啦

**Comment by YC0315** [comment]
> 思达，我有问题，我只输入rgb图片，用了如下几张图片，为什么我用自己拍摄的环境，猫咪图片不可以预测，加了标定板就好一些，而blender融合的图片没有标定板就能预测呢？是否验证需要标定板呢？相机标定如何确定得到的内参是准确呢？（我的做法：7*9的标定纸固定不变，相机不同角度横着拍摄了20张照片，代码执行出的相机内参）
> 
> 1. LINMEMOD的任意一张图片。可预测
>    ![Figure_444](https://user-images.githubusercontent.com/51335035/68195694-3fdb0e80-fff2-11e9-8f14-4cef0a6fae8f.png)
> 
> 2.blender小猫猫的一张图片。可预测 ![Figure_8879](https://user-images.githubusercontent.com/51335035/68195733-508b8480-fff2-11e9-81f9-27a87bb08815.png)
> 
> 3.电脑摄像头拍的一张照片，不可以，坐标点很乱。（有用到自己标定的内参矩阵，但是不知道标定的如何才算准确） ![Figure_666](https://user-images.githubusercontent.com/51335035/68195767-68fb9f00-fff2-11e9-8f3d-0704feda1fc3.png)
> 
> 4.把小猫猫放在标定纸上，再用相机拍摄，然后用了我的相机内参和原代码里的。 （1）用我的相机内参 ![Figure_000](https://user-images.githubusercontent.com/51335035/68196187-2a1a1900-fff3-11e9-8b63-a54e422677e9.png)
> 
> （2）用原代码相机内参 ![Figure_8](https://user-images.githubusercontent.com/51335035/68195794-7c0e6f00-fff2-11e9-986a-39ac1c3c5b38.png)

您好，我想请教您个问题，就是我也是使用自己的相机拍摄的3D打印的cat模型的10张图片，然后用自己训练好的模型去生成包围框，但是所有的包围框都是错的，而且都是一个状态。我想问下如果检测自己拍摄的图片上物体位姿，是不是只需要物体图片，相机内参就可以了呢？请问3D关键点坐标怎么得到呢？

**Comment by bckaiwang** [comment]
@emojiLee 请问meta文件是怎么获得的呀？

**Comment by Harrydsk** [comment]
> 嗯嗯，谢谢思达，真厉害

你好啊！ 怎么先把图片中的物体crop出来
